File: 2023/scala/src/main/scala/day02.scala
package day02

import locations.Directory.currentDir
import inputs.Input.loadFileSync

val input = """Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green"""

@main def part1: Unit =
  println(s"The solution is ${part1(loadInput())}")

@main def part2: Unit =
  println(s"The solution is ${part2(loadInput())}")

def loadInput(): String = loadFileSync(s"$currentDir/../../../../input/day02")

def part1(input: String): String =

  val totalRed = 12
  val totalGreen = 13
  val totalBlue = 14

  input.linesIterator
    .map {
      case s"Game $id: $sets" => {
        val (possibleId: Int, green: Int, red: Int, blue: Int) =
          sets.split("; ").foldLeft(id.toInt, 0, 0, 0) {
            case ((id, green, red, blue), set1) =>
              set1.split(", ").foldLeft(id, green, red, blue) {
                case ((id, green, red, blue), s"$n green") =>
                  (id, if n.toInt > green then n.toInt else green, red, blue)

                case ((id, green, red, blue), s"$n red") =>
                  (id, green, if n.toInt > red then n.toInt else red, blue)

                case ((id, green, red, blue), s"$n blue") =>
                  (id, green, red, if n.toInt > blue then n.toInt else blue)

              }
          }
        if green <= totalGreen && blue <= totalBlue && red <= totalRed then {
          possibleId
        } else {
          0
        }
      }
    }
    .sum
    .toString()

def part2(input: String): String =
  input.linesIterator
    .map {
      case s"Game $id: $sets" => {
        val (_: Int, green: Int, red: Int, blue: Int) =
          sets.split("; ").foldLeft(id.toInt, 0, 0, 0) {
            case ((id, green, red, blue), set1) =>
              set1.split(", ").foldLeft(id, green, red, blue) {
                case ((id, green, red, blue), s"$n green") =>
                  (id, if n.toInt > green then n.toInt else green, red, blue)

                case ((id, green, red, blue), s"$n red") =>
                  (id, green, if n.toInt > red then n.toInt else red, blue)

                case ((id, green, red, blue), s"$n blue") =>
                  (id, green, red, if n.toInt > blue then n.toInt else blue)

              }
          }
        green * red * blue
      }
    }
    .sum
    .toString()

@main def part12: Unit =

  val input = loadInput()

  val totalRed = 12
  val totalGreen = 13
  val totalBlue = 14

  val (sumOfIds, sumOfPower) = input.linesIterator
    .foldLeft(0, 0) {
      case ((sumIds, sumOfPower), s"Game $id: $sets") => {
        val (possibleId: Int, green: Int, red: Int, blue: Int) =
          sets.split("; ").foldLeft(id.toInt, 0, 0, 0) {
            case ((id, green, red, blue), set1) =>
              set1.split(", ").foldLeft(id, green, red, blue) {
                case ((id, green, red, blue), s"$n green") =>
                  (id, if n.toInt > green then n.toInt else green, red, blue)

                case ((id, green, red, blue), s"$n red") =>
                  (id, green, if n.toInt > red then n.toInt else red, blue)

                case ((id, green, red, blue), s"$n blue") =>
                  (id, green, red, if n.toInt > blue then n.toInt else blue)

              }
          }
        if green <= totalGreen && blue <= totalBlue && red <= totalRed then {
          (sumIds + possibleId, green * red * blue + sumOfPower)
        } else {
          (sumIds, green * red * blue + sumOfPower)
        }
      }
    }

  println(s"Solution for part1 is $sumOfIds")
  println(s"Solution for part2 is $sumOfPower")

File: 2023/haskell/src/Day02.hs
module Day02 (part1, part2) where

import Data.List.Split (splitOn)
import qualified Data.Map as M
import Data.Maybe (fromMaybe)

data Color = Red | Green | Blue deriving (Show, Eq, Ord)

newtype GameSet = GameSet [(Int, Color)] deriving (Show)

data Game = Game Int [GameSet] deriving (Show)

getColor :: String -> Color
getColor "green" = Green
getColor "red" = Red
getColor "blue" = Blue
getColor _ = error "Not a valid color!"

createMap :: Game -> M.Map Color Int
createMap (Game _ sets) =
  foldl (\m (GameSet xs) -> foldl (\m1 (count, color) -> M.insertWith max color count m1) m xs) M.empty sets

notFewer :: (Int, Int, Int) -> M.Map Color Int -> Bool
notFewer (red, green, blue) m = do
  foldl
    ( \ok (color, threshold) -> case M.lookup color m of
        Just r -> ok && r <= threshold
        Nothing -> False
    )
    True
    [(Red, red), (Green, green), (Blue, blue)]

powerOfTheSets :: M.Map Color Int -> Int
powerOfTheSets m =
  product $ map (\color -> fromMaybe 0 (M.lookup color m)) [Red, Green, Blue]

createGame :: String -> Game
createGame s = do
  Game gameId sets
  where
    (gameIdStr, rest) = case span (/= ':') s of
      ('G' : 'a' : 'm' : 'e' : ' ' : gameIdStr1, ':' : ' ' : rest1) -> (gameIdStr1, rest1)
      _ -> error "Unexpected input format!"
    gameId = (read :: String -> Int) gameIdStr
    xs = splitOn "; " rest
    sets = map (GameSet . map ((\ws -> ((read :: String -> Int) (head ws), getColor (ws !! 1))) . words) . splitOn ", ") xs

part1 :: String -> IO ()
part1 input = do
  let games = map createGame $ lines input
  print $ sum $ map (\(Game gameId _) -> gameId) $ filter (notFewer (12, 13, 14) . createMap) games

part2 :: String -> IO ()
part2 input = do
  let games = map createGame $ lines input
  print $ sum $ map (powerOfTheSets . createMap) games